### PR TITLE
Optimize modules loading.

### PR DESCRIPTION
**Pull request by jtkech**
- Features loading was done through the ctor of `ShellContainerFactory`, here it is normally triggered by `ShellHost` just before loading shell settings.

- At some point, each module was loaded inside the `Parallel.ForEach()` of `ExtensionManager`, but the module was loaded inside a lock block and it would have been better to load it outside and just update the related non concurrent dico inside the lock block.

- Anyway, now modules can only be loaded in one pass and this is done before the `Parallel.ForEach()` of `ExtensionManager`. But that's ok, so here the suggestion is just to load modules in parallel. The workload is not so heavy but enough to benefit from doing it in parallel when there are many modules.

### CODE REVIEW

File: src/OrchardCore/OrchardCore.Abstractions/Modules/ModularApplicationContext.cs [hunk 1]
```diff
@@ -48,12 +49,17 @@ private void EnsureInitialized()
...
 
             foreach (var provider in _moduleNamesProviders)
             {
-                modules.AddRange(provider.GetModuleNames().Select(name => new Module(name, name == _environment.ApplicationName)));
+                var names = provider.GetModuleNames().Distinct();
+
+                Parallel.ForEach(names, new ParallelOptions { MaxDegreeOfParallelism = 8 }, (name) =>
+                {
+                    modules.Add(new Module(name, name == _environment.ApplicationName));
+                });
```
Comment (Jetski5822) [inline src/OrchardCore/OrchardCore.Abstractions/Modules/ModularApplicationContext.cs]: Would it be better to stick using Linq, but do AsParrellel() ?

File: src/OrchardCore/OrchardCore.Abstractions/Modules/ModularApplicationContext.cs [hunk 2]
```diff
@@ -48,12 +49,17 @@ private void EnsureInitialized()
...
             modules.Add(new Module(_environment.ApplicationName, true));
 
             foreach (var provider in _moduleNamesProviders)
             {
-                modules.AddRange(provider.GetModuleNames().Select(name => new Module(name, name == _environment.ApplicationName)));
+                var names = provider.GetModuleNames().Distinct();
+
+                Parallel.ForEach(names, new ParallelOptions { MaxDegreeOfParallelism = 8 }, (name) =>
+                {
+                    modules.Add(new Module(name, name == _environment.ApplicationName));
```
Comment (Jetski5822) [inline src/OrchardCore/OrchardCore.Abstractions/Modules/ModularApplicationContext.cs]: I am wondering... should name and application name be case insensitive? :/

Comment (Jetski5822) [reply src/OrchardCore/OrchardCore.Abstractions/Modules/ModularApplicationContext.cs]: I actually think this whole section is kinda wrong... shouldnt it be more like this...

```
	var modules = _moduleNamesProviders
		.SelectMany(x => x.GetModuleNames())
		.Distinct()
		.AsParallel()
		.Select(name => new Module(name, name == _environment.ApplicationName));
```

**Comment by jtkech** [comment]
@Jetski5822 thanks for your review

- About names it's because in some env assembly file names may not be case insensitive.

- I didn't used `AsParallel` just because, as i have tried, a `ConcurrentBag` and `Parallel.ForEach` was faster.

- I thought about using all providers in one pass but i kept the existing foreach loop just to respect what we have in `Application.cs`.

      private readonly Dictionary<string, List<Module>> _modulesByName;
      ...
      _modulesByName = _modules.GroupBy(x => x.Name).ToDictionary(x => x.Key, x => x.ToList());
      ...
        public Module GetModule(string name)
        {
            if (!_modulesByName.TryGetValue(name, out var modules) || modules.Count == 0)
            {
                return new Module(string.Empty);
            }

            return modules[modules.Count - 1];
        }

    `_modulesByName` is not a dictionary of modules, it is a dictionary of `List<Module>`.

    But here you're right, because at the end for a given name only one module is provided, i will change it to use all providers in one pass, return distinct names, and in `Application.cs` just use a dictionary of modules.

**Pull request merged**